File: src/scope.rs
use std::any::Any;
use std::collections::HashMap;

pub enum AllOrList<T> {
    All,
    List(Vec<T>),
}

impl<T> AllOrList<T> {
    fn push(&mut self, v: T) {
        match self {
            AllOrList::All => {
                *self = AllOrList::List(vec![v]);
            }
            AllOrList::List(l) => {
                l.push(v);
            }
        }
    }
}

impl<T> Default for AllOrList<T> {
    fn default() -> Self {
        AllOrList::All
    }
}

pub type Field = AllOrList<String>;
pub type PredicateFunction = Fn(HashMap<String, &Any>) -> bool;
pub type Predicate = AllOrList<Box<PredicateFunction>>;

pub struct Scope {
    provider: String,
    path: String,
    predicates: Predicate,
    fields: Field,
}

impl Scope {
    pub fn new(provider: String, path: String) -> Scope {
        Scope {
            provider: provider,
            path: path,
            predicates: Default::default(),
            fields: Default::default(),
        }
    }

    pub fn add_filter(mut self, predicate: Box<PredicateFunction>) -> Scope {
        self.predicates.push(predicate);
        self
    }

    pub fn add_field(mut self, field: String) -> Scope {
        self.fields.push(field);
        self
    }
}

File: processor/base/js/src/scope.js
const processorRouter = require('../../../providers/all.js');

let predicateByOps = {
    "=":   function(e, field, val) { return e[field] === val; },
    "!=":  function(e, field, val) { return e[field] !== val; },
    ">":   function(e, field, val) { return e[field] > val; },
    ">=":  function(e, field, val) { return e[field] >= val; },
    "<":   function(e, field, val) { return e[field] < val; },
    "<=":  function(e, field, val) { return e[field] <= val; },
    "has": function(e, field) { return field in e },
};

class Scope {
    constructor(scope) {
        this.provider = scope.provider;
        this.path = scope.path;
        this.predicates = scope.predicates || [];
        this.fields = scope.fields || null;
    }

    filter(col) {
        return col.filter(this.match.bind(this)).map(this.mapFields.bind(this));
    }


    match(e) {
        for (const pred of this.predicates) {
            const op = pred[0];
            const predFunc = predicateByOps[op];
            if (!predFunc) {
                throw "unknown predicate operator: " + op;
            }

            if (!predFunc(e, ...pred.slice(1))) {
                return false;
            }
        }

        return true;
    }

    mapFields(e) {
        if (this.fields == null) {
            return e;
        }

        let o = {};
        for (const i in this.fields) {
            const field = this.fields[i];
            o[field] = e[field];
        }

        return o;
    }

    hasField(field) {
        if (this.fields == null) {
            return true;
        }

        return this.fields.indexOf(field) != -1;
    }

    describe() {
        const r = {
            provider: this.provider,
            path: this.path,
            predicates: this.predicates,
            fields: this.fields,
        };

        const proc = processorRouter.get(this.provider);
        if (proc == null) {
            return r;
        }

        r.providerDesc = proc.name;

        const pathDesc = proc.descByPath[this.path];
        if (pathDesc) {
            r.pathDesc = pathDesc;
        }

        // XXX add predicates
        // XXX add fields

        return r;
    }
}

module.exports.Scope = Scope;
